iina/PlayerCore.swift

```diff
@@ -642,10 +651,169 @@ class PlayerCore: NSObject {
+
+    // search subs
+    let fm = FileManager.default
+    let searchOptions: FileManager.DirectoryEnumerationOptions = [.skipsHiddenFiles, .skipsPackageDescendants, .skipsSubdirectoryDescendants]
+    let subExts = Utility.supportedFileExt[.sub]!
+    var subDirs: [URL] = []
+
+    // search subs in current directory
+    guard let files = try? fm.contentsOfDirectory(at: folder, includingPropertiesForKeys: nil, options: searchOptions) else { return }
+
+    // search subs in other directories
+    let rawUserDefinedSearchPaths = ud.string(forKey: Preference.Key.subAutoLoadSearchPath) ?? "./*"
+    let userDefinedSearchPaths = rawUserDefinedSearchPaths.components(separatedBy: ":").filter { !$0.isEmpty }
+    for path in userDefinedSearchPaths {
+      var p = path
+      // handle absolute paths
+      if path.hasPrefix("/") {
+        subDirs.append(URL(fileURLWithPath: path, isDirectory: true))
+      }
+      if path.hasPrefix("~") {
+        subDirs.append(URL(fileURLWithPath: NSString(string: path).expandingTildeInPath, isDirectory: true))
+      }
+      // handle wildcards
+      if path.hasSuffix("/") { p.deleteLast(1) }
+      if path.hasSuffix("/*") {  // only check wildcard at the end
+        p.deleteLast(2)
+        let pathURL = folder.appendingPathComponent(p, isDirectory: true)
+        // append all sub dirs
+        if let contents = try? fm.contentsOfDirectory(at: pathURL, includingPropertiesForKeys: [.isDirectoryKey], options: [.skipsHiddenFiles, .skipsPackageDescendants]) {
+          if #available(OSX 10.11, *) {
+            subDirs.append(contentsOf: contents.filter { $0.hasDirectoryPath })
+          } else {
+            subDirs.append(contentsOf: contents.filter { (try? $0.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory ?? false })
+          }
+        }
+      } else {
+        subDirs.append(folder.appendingPathComponent(p, isDirectory: true))
+      }
+    }
+
+    // group by extension
+    var groups: [MPVTrack.TrackType: [FileInfo]] = [.video: [], .audio: [], .sub: []]
+    let allTypes: [MPVTrack.TrackType] = [.video, .audio, .sub]
+    for file in files {
+      let fileInfo = FileInfo(file)
+      guard let mediaType = allTypes.first(where: { Utility.supportedFileExt[$0]!.contains(fileInfo.ext) }) else { continue }
+      groups[mediaType]!.append(fileInfo)
+    }
+
+    // natural sort
+    groups[.video]!.sort { file1, file2 -> Bool in
+      return file1.segments.lexicographicallyPrecedes(file2.segments) { a, b in
+        if let inta = Int(a), let intb = Int(b) {
+          return inta < intb
+        } else {
+          return a < b
+        }
+      }
+    }
+
+    // get all possible sub files
+    var subtitles = groups[.sub]!
+    for subDir in subDirs {
+      if let contents = try? fm.contentsOfDirectory(at: subDir, includingPropertiesForKeys: nil, options: searchOptions) {
+        subtitles.append(contentsOf: contents.filter { subExts.contains($0.pathExtension) }.map { FileInfo($0) })
```

Comment:
`flatMap` will do this in a single passthrough, instead of the two that `filter` and `map` require.